runtime/src/bank_forks.rs

```diff
@@ -86,8 +86,55 @@ impl Index<u64> for BankForks {
 
 impl BankForks {
     pub fn new_rw_arc(bank: Bank) -> Arc<RwLock<Self>> {
+        let bank = Arc::new(bank);
         let root = bank.slot();
-        Self::new_from_banks(&[Arc::new(bank)], root)
+
+        let mut banks = HashMap::new();
+        banks.insert(
+            bank.slot(),
+            BankWithScheduler::new_without_scheduler(bank.clone()),
+        );
+        let parents = bank.parents();
+        for parent in parents {
+            if banks
+                .insert(
+                    parent.slot(),
+                    BankWithScheduler::new_without_scheduler(parent.clone()),
+                )
+                .is_some()
+            {
+                // All ancestors have already been inserted by another fork
+                break;
+            }
+        }
+
+        let mut descendants = HashMap::<_, HashSet<_>>::new();
+        for (slot, bank) in &banks {
+            descendants.entry(*slot).or_default();
+            for parent in bank.proper_ancestors() {
+                descendants.entry(parent).or_default().insert(*slot);
+            }
+        }
+        let bank_forks = Arc::new(RwLock::new(Self {
+            root: Arc::new(AtomicSlot::new(root)),
+            banks,
+            descendants,
+            snapshot_config: None,
+            accounts_hash_interval_slots: std::u64::MAX,
+            last_accounts_hash_slot: root,
+            in_vote_only_mode: Arc::new(AtomicBool::new(false)),
+            highest_slot_at_startup: 0,
+            scheduler_pool: None,
+        }));
+
+        for bank in bank_forks.read().unwrap().banks.values() {
+            bank.loaded_programs_cache
+                .write()
+                .unwrap()
+                .set_fork_graph(bank_forks.clone());
+        }
```

Comment:
Same as above - think we should operate on that known slot directly instead of looping ?

Reply:
they should all have the same cache anyway, so I think we're good to only call this on the root bank. Is that correct @pgarg66?

Reply:
If the banks were created using `new_from_parent` constructor, then all of them will clone their parent's cache instance. i.e. the first/root bank was created from snapshot (or was bank0), and the rest others were created from that bank.

Reply:
Got it. Well given that `BankForks` can only have exactly one bank at construction, I think we should explicitly avoid the looping. If someone create a `Bank` without using `Bank::new_from_parent()`, that is programmer error and we can't prevent that with any logic in `BankForks`.**

**Technically, we could enforce this by calling `set_fork_graph()` for every `Bank` that we push into `BankForks`, but I'm pretty confident we wouldn't want to incur this overhead.